Model error as object

lulalala opened the pull request:
### Summary

Add `ActiveModel::Error` class for encapsulating a single error. It handles message generation and details access. 
Utilize this in `Errors` class, changing it from a hash based interface, to an array of `Error` objects. Add more flexible query methods like `where`.

### Introduction

ActiveModel#errors interface is currently not very object oriented. For some complex use cases, this design made it a bit tedious to use. I feel these issues can be remedied by encapsulating each individual error as an object.

Last year I implemented AdequateErrors gem to do this, and it solved many of my problems. I later found out that back in 2016, @eprothro suggested the [same idea too](https://groups.google.com/forum/#!topic/rubyonrails-core/dRxdSeV4JQE) on the core mailinglist, and Rafael was positive about this. So I made this PR.

I've made it to respect Rails’ deprecation policy. As this contains breaking changes, can this go straight into Rails 6.0?

I will fix other specs/add doc/performance tuning once interface is finalized.

### Benefits

#### More flexible query interface such as `where`

It’s easy to query one particular object using the `where` clause.

```ruby
model.errors.where(:name, :foo, bar: 3).first
```

`delete`, `add`, `added?`, `where` all share the same method signature (attribute, type, options). So we are able to delete specific errors now:

```ruby
model.errors.delete(:name, :too_powerful, level: 9000)
```

#### Testing is more precise and flexible

We can now test if “foo” error exists, regardless of its options hash.

```ruby
model.errors.add(:name, :too_powerful, level: 9000)

model.errors.added?(:name, :too_powerful, level: 9000) # returns true
model.errors.added?(:name, :too_powerful) # will be false in the past, but be true now.
```

In the past, added? works by re-render the message and compare it against current message. Therefore if level is not provided, it will return false. In the PR, `added?` only compare Error's attributes against what's provided, so it can be more general or more specific depending on the needs.

#### Get message of corresponding details of one particular error

If you saw that name attribute has two `foo_error` and one `bar_error`, e.g.:

```ruby
# model.errors.details
{:name=>[{error: :foo_error, count: 1}, {error: :bar_error}, {error: :foo_error, count: 3}]}
```

How do you back track the message on the third particular error? With the current implementation, we have to resort to using array indexes:

```ruby
model.errors.messages[:name][2]
```

Or we can call `generate_message` to recreate a message from the details, but that's also tedious.

With OO, we won't have this problem. Error is represented as an object, message and details are its attributes, so accessing those are straightforward:

```ruby
e = model.errors.where(:name, :foo_error).last
e.full_message
e.options # similar to details, where meta informations such as `:count` is stored.
```

#### Lazily evaluating message for internationalization

[@morgoth](https://github.com/morgoth) mentioned this issue that when you're adding error, it's translated right away.

```ruby
# actual:
I18n.with_locale(:pl) { user.error.full_messages } # => outputs EN errors always

# expecting:
I18n.with_locale(:pl) { user.error.full_messages } # => outputs PL errors
I18n.with_locale(:pt) { user.error.full_messages } # => outputs PT errors
```

This PR addresses this by lazily evaluating messages only when `message` is called.

#### Opens possibility to advanced modding

Once errors are objects, it’s easy to add functionality on top of them. We can have custom methods to disable global attribute prefix on error’s full messages.

### List of API changes

`[]`
unchanged, deprecated (use `messages_for` instead)

`as_json`, `blank?`, `clear`, `count`, `empty?`
unchanged

`add`
unchanged

`added?`
mostly unchanged, for the one change see "Testing is more precise and flexible"

`delete`
extended, so we can give more specific condition such as error type or options.

`each`
If we use `each{|attr,msgs|}` then it behaves the same as before. Deprecated
If we use `each{|error|}` then it loops through Error array.

`full_message`
deprecated as it is no longer needed.
unchanged ~~message is generated in Error.~~

`full_messages`, `full_messages_for`
unchanged

`messages_for`
new, to replace deprecated `[]`

`where`
new, query for error objects
`generate_message`
deprecated ~~Part of Error as message is generated there.~~

`has_key?` `key?`, `keys`
deprecated

`include?`, `size`, `to_hash`, `to_xml`, `to_a`
unchanged

`values` 
deprecated

`import`
new, imports one error as a nested error. Useful in Form Object and nested attribute error importing.

`messages`, `details`
unchanged 

### Some questions I have

~~I am not sure what's the policy for marshal across versions. `marshal_dump` and `marshal_load` are implemented, but do they have to support marshaling across Rails versions?~~

This has been done.

Can we deprecate `to_xml`? I bet is rarely used, and can exists as a gem.

lulalala committed:
Add ActiveModel::Error and NestedError

Add initialize_dup to deep dup.

Move proc eval and flexible message position out to Errors,
because proc eval is needed for Errors#added? and Errors#delete

lulalala committed:
Change errors

Allow `each` to behave in new way if block arity is 1

Ensure dumped marshal from Rails 5 can be loaded

Make errors compatible with marshal and YAML dumps from previous versions of Rails

Add deprecation warnings

Ensure each behave like the past, sorted by attribute

pixeltrix commented:
@lulalala is there a reason why you can't emulate the hash-based API?

lulalala commented:
@pixeltrix for the hash like APIs, my thoughts are:

1. For enumerable methods such as `each`, I choose to forward them to the error array. Keeping it to enumerate as hash seems to defeat the purpose.
2. For `[]` and `to_hash`, they are emulated.
3. For `values`, I removed it. I feel it was there just so `errors` feels like a hash. It is not useful because having an array of partial error message without knowing each message's respective attribute is useless.

rafaelfranca commented:
Everything that was removed needs to be deprecated first. We don't make breaking changes without deprecation, even in major versions.

> marshal_dump and marshal_load are implemented, but do they have to support marshaling across Rails versions?

Yes, we try very hard to make possible that an old Rails version can read data from the new version and also the opposite. This make possible for applications to slowly rollout new versions to production.

rafaelfranca commented on activemodel/lib/active_model/error.rb:
```diff
@@ -0,0 +1,120 @@
+
+module ActiveModel
+  # Represents one single error
+  # @!attribute [r] base
```
We don't use yarn, we use rdoc, so we need to follow the Rails documentation guideline.

lulalala commented:
> Everything that was removed needs to be deprecated first. We don't make breaking changes without deprecation, even in major versions.

Cool. I'll add those back then. How about semantic change such as `each`?

> Yes, we try very hard to make possible that an old Rails version can read data from the new version and also the opposite. This make possible for applications to slowly rollout new versions to production.

Got it, but I still have a question. When `marshal_load` was added (SHA b3dfd7d16cc37dcaf4298fe42e69a56ab3c6b00e), I don't see it tried to accommodate Rails 3 or 4, or am I wrong?

rafaelfranca commented:
> Got it, but I still have a question. When marshal_load was added (SHA b3dfd7d), I don't see it tried to accommodate Rails 3 or 4, or am I wrong?

Usually we only think about backward compatibility between two close release. In that case 5.1 and 5.2. Right now it would be 6.0 and 5.2

rafaelfranca commented:
> How about semantic change such as each?

Can we introduce a new method with the new semantic and keep the old method with the previous semantic but with deprecation?

lulalala commented:
> Can we introduce a new method with the new semantic and keep the old method with the previous semantic but with deprecation?

I realized there is a better solution. I check arity of the block passed to `each`. If arity is 1, behave the new way. We can put deprecation notice when the old way is triggered. Thoughts?
https://github.com/rails/rails/pull/32313/files#diff-fdcf8b65b5fb954372c6fe1ddf284c78R196

rafaelfranca commented:
That is a good eay. 👍

lulalala committed:
Add convenience method group_by_attribute

Many operations need grouping of errors by attributes, e.g. ActiveRecord::AutosaveAssociation#association_valid?

Refactor other methods using group_by_attribute

lulalala committed:
Use errors#import instead of manipulating errors/details hash

lulalala committed:
Fix misalignment caused by SHA eebb9ddf9ba559a510975c486fe59a4edc9da97d

lulalala commented:
I have added back existing methods, some with deprecation warning (free to discuss about which to deprecate). The `each` now should be compatible with old hash block.

The YAML and marshal dumps are also made compatible with past versions. I added tests for each.

I am starting to look into other tests which broke, and I learned a lot about what interface are needed. For example I added `group_by_attribute` because it is suitable for many occasions.

I'll update in the future once all tests are fixed :)

lulalala committed:
Raise deprecation for calling `[:f] = 'b'` or `[:f] << 'b'`

Revert some tests to ensure back compatibility

lulalala committed:
Backward compatibility for errors.collect/select etc.

All enumerable methods must go through the `each` so it retain old hash behavior.
Revert this after Rails 6.1 in order to speed up enumerable methods.

lulalala committed:
Add a transitional method `objects`, for accessing the array directly.

This is because we try to accommodate old hash behavior, so `first` and `last` now does not return Error object.

lulalala committed:
Add messages_for

lulalala committed:
Allow errors to remove duplicates, and ensure cyclic associations w/ autosave duplicate errors can be removed

See SHA 7550f0a016ee6647aaa76c0c0ae30bebc3867288

lulalala committed:
String override options in #import to convert to symbol

lulalala committed:
Nested attribute error's attribute name to use different key:

To keep the same as SHA dcafe995bfe51e53dd04607956be9b54073e9cb6

lulalala commented:
I think I have fixed almost all tests (the remaining few seems to be sporadic).

I discovered there are many cases of `errors[:foo] << 'bar'`. We may be able to put a deprecation warning on that, but would involves having a custom array class which overrides `<<` just for that purpose. I hope that's not needed.

Another behavior change is that details will not auto populate with empty array if it is access with a missing key.

If the deprecation proposal here is finalized, I'll start adding code so tests won't have deprecation warning all over the place.

lulalala commented:
Hi @tenderlove, we have met in Ruby Kaigi, and you mentioned that for emulated message arrays, I should freeze those so attempts to append to them would fail. This is now added. Cheers! (Sorry I forgot who was sitting next to you who said I can also ping him as well)

lulalala commented:
Hi @Larochelle, I see that you made a few changes lately to errors.rb. Just wondering if you plan to make further changes, and whether you are interested in this MR?

I am slowly trying to resolve the conflict, but I am a little unsure if my resolve is what you intended here. Would you be free to check my commits? Thanks.

Larochelle commented:
> Would you be interested in working with me on this together?

Unfortunately, I don't have any bandwidth for the next couple months.  
Rafael knows all the context I have, so I trust the solution will be great once he 👍

lulalala commented:
Hi @rafaelfranca, I could not reach you via email/twitter, so want to ask if you can review this please? Thanks!

tenderlove commented on activemodel/test/cases/errors_test.rb:
```diff
@@ -31,30 +31,29 @@ def self.lookup_ancestors
   def test_include?
     errors = ActiveModel::Errors.new(self)
-    errors[:foo] << "omg"
+    errors.add(:foo, "omg")
```
Is it possible to wrap `errors[:foo] << "omg"` with `assert_deprecated { }` and have the test still pass?

lulalala replied:
@tenderlove Yes got it. However this particular case was one of the incompatibility that can' really be solved. This is an emulated hash and you mentioned about placing a freeze to it.

However it is possible if I return a subclass of Hash, and override `<<`. If you want this I will try implementing it.

tenderlove reviewed (changes requested):
Hi,

I like this change, but the diff is very large.  I'd feel more comfortable merging if the changes in `errors_test.rb` had `assert_deprecated { }` around existing method calls.  It's fine if we have duplicate tests, one that uses `m.add(x, y)` and one that does `m.errors[x] << y`.  We can remove the duplicate tests later.  Wrapping the existing calls with `assert_deprecated { }` would make me feel much better about merging this.

Thank you!

matthewd commented:
> Sorry I forgot who was sitting next to you who said I can also ping him as well

👋🏻:sweat_smile: 

Hmm.. when I suggested just freezing the array, I didn't realise `<<` was currently such a prominent way of adding errors: even the guide uses it repeatedly (and says "An equivalent to errors#add is to use <<").

Given that, I think we will need to do the ugly option, and create an array wrapper... something like:

```ruby
class ErrorArray < SimpleDelegator
  def initialize(content, &append)
    @append = append
    super(content)
  end

  def <<(message)
    __setobj__ @append.call(message)
    self
  end
end

def [](attribute)
  ErrorArray.new(messages_for(attribute)) do |message|
    ActiveSupport::Deprecation.warn("plz no")
    add attribute, message
    messages_for(attribute)
  end
end
```

Note this:
* still has `messages_for` (which is unchanged) freeze the inner array, so other less common array methods will still fail in an obvious way 
* updates itself after `<<` so people can see their newly-added error
* only returns the wrapper from the deprecated `[]` method: people using the new API know it returns a frozen array

It does not reflect changes made directly via `errors.add` -- it seems unreasonable for a caller to do that while holding this separate array and rely on this to change, whereas a write+read on a single object seems more likely.

(Sorry, I wanted to think out how terrible it would be, and the easiest way to do that was to write it.)

lulalala committed:
Add deprecation to slice!

lulalala commented:
Thanks @matthewd @rafaelfranca I've covered the case of `errors[:foo] << 'bar'` and `errors[:foo] = 'bar'`

The rabbit hole can go much deeper, but I hope this is enough.

Due to the fact that the messages array is freezed, we had to regenerate a new one everytime if `[] <<` is used. If performance is a concern, then we have to remove the freeze. What's your opinions on that?

---

There is one behavior change I am against doing back-compatible implementation is: 

Say my error details only contains `:title` error, if I still do `details[:foo]`, in the past it would give me `[]`. Now it would return nil. The reason it returned `[]` was solely because it allows users to do `<<` to it. So can I not emulate this behavior in the new implementation?

matthewd commented:
Do we need to handle the `to_hash` case? Is there documentation / wide usage of that pattern? I was only really concerned about the `errors[:foo]` array.

Per my example, I don't think the new `messages_for` method has any reason to return the wrapper.

That would also fix `__setobj__ @errors[@attribute]`, which currently sounds like it would 1) trigger an extra deprecation warning, and 2) leave the wrapper pointing at another wrapper instance.

lulalala commented on activemodel/test/cases/errors_test.rb:
```diff
@@ -86,47 +85,55 @@ def test_no_key
     errors = ActiveModel::Errors.new(self)
-    errors.messages[:foo] = "omg"
-    errors.messages[:baz] = "zomg"
+    assert_deprecated { errors.messages[:foo] = "omg" }
```
@matthewd About the to_hash's `[] =` case, it was probably not mentioned in the doc, but this spec provides direct usage. So unless we remove this spec, we had to tempter with `to_hash`.

lulalala committed:
Split messages and to_hash

Fix double wrapping issue
Revert messages_for wrapping. It's a new method so no need to put
deprecation warnings.

lulalala committed:
Assert deprecation

lulalala committed:
Fix spec

lulalala commented:
@rafaelfranca @tenderlove just wondering, would it be possible to temporarily request new pull requests to be based on this branch instead of master?

Resolving conflicts repeatedly has been quite time consuming. I had to reinvest a few hours to re-understand and re-implement every newly merged PRs. This has happened 5~6 times now.

tenderlove commented on activemodel/test/cases/error_test.rb:
```diff
@@ -0,0 +1,196 @@
+    assert_equal false, subject.match?(:mineral, :too_coarse)
+    assert_equal true, subject.match?(:mineral, :not_enough)
+    assert_equal true, subject.match?(:mineral, :not_enough, count: 2)
+    assert_equal false, subject.match?(:mineral, :not_enough, count: 1)
```
Can we change these to `assert` and `assert_not`?  I don't think asserting a specific boolean return value is necessary (maybe we want to return `nil` or some truthy object later).

tenderlove commented on activemodel/lib/active_model/error.rb:
```diff
@@ -0,0 +1,197 @@
+
+    def hash
+      attributes_for_hash.hash
+    end
```
Is implementing all of these necessary?  I can't imagine a use case for comparing semantically comparing error objects.

tenderlove commented on activemodel/lib/active_model/error.rb:
```diff
@@ -0,0 +1,197 @@
+    def ==(other)
+      attributes_for_hash == other.attributes_for_hash
+    end
+    alias eql? ==
```
This probably needs to check the type, or a `respond_to?`.  I think code like this might raise an exception:

```ruby
if error_object == nil
  ...
end
```

I know it seems like a contrived example, but nobody would expect `==` to raise ever.

lulalala replied:
Ah, I would have never thought of that!

tenderlove reviewed (commented):
Implementing the equality / hash operators seems suspicious to me.  I'd like it if they were removed.  Other than that, this PR looks fine to me.

lulalala committed:
Fix equality comparison raising error bug

lulalala commented:
@tenderlove @rafaelfranca Thanks and I've made the changes. Could you review? Thanks

tenderlove reviewed (approved):
I'm good with this.  @rafaelfranca do you mind taking one more look?

matthewd commented:
`Errors#[]` looks very expensive; I think it's worth changing it to:

```ruby
def [](attribute)
  DeprecationHandlingMessageArray.new(messages_for(attribute), self, attribute)
end
```

---

I also think it's important that both `DeprecationHandlingMessageHash` and `DeprecationHandlingMessageArray` freeze their target, to protect against other mutation methods. (So `DeprecationHandlingMessageHash#[]=` will have to use the `__setobj__` technique like `DeprecationHandlingMessageArray#<<`.)

---

Finally, I wonder whether it would be practical to cache the `DeprecationHandlingMessageHash` inside `Errors#messages` (and then clear it whenever we change `@errors`) -- it's not a showstopper, but seems a bit unfortunate to change previously-very-cheap `errors.messages[:a], errors.messages[:b], etc` into a sequence of expensive hash construction.

lulalala commented:
@matthewd  thanks, I've done the first suggestion.

About the freezing the hash, one thing is stopping it. The hash has a `default_proc`, so we can get an empty array when we access some attribute which were previously not in the hash.

For the performance refactor, I plan to do it after the merge, otherwise some other MR might cause some conflict again, requiring me to rewire all the internals.

matthewd commented:
> The hash has a default_proc, so we can get an empty array when we access some attribute which were previously not in the hash.

Hmm.. I think we can solve that using `__setobj__` too:

```ruby
content.default_proc = proc do |hash, attribute|
  hash = hash.dup
  hash[attribute] = DeprecationHandlingMessageArray.new([], @errors, attribute)
  __setobj__ hash.freeze
  hash[attribute]
end
```

Really, as we're not caching the hash, this would probably be fine too:

```ruby
content.default_proc = proc do |hash, attribute|
  DeprecationHandlingMessageArray.new([], @errors, attribute)
end
```

Either way, the freeze in `DeprecationHandlingMessageArray` should be safe, so let's at least get that one in there.

lulalala committed:
Freeze DeprecationHandling array and hash

lulalala commented:
@matthewd ah I didn't thought of that. Thanks and updated.

lulalala commented:
@rafaelfranca WDYT? Thanks.

tyok commented:
Sorry for commenting this late, but how about renaming `where` to `select`? I see you even implemented it by calling `@errors.select`.

Going further, implementing the inverse is as simple as implementing `reject` (instead of `where.not`, which seems like a lot of hassle).

lulalala commented:
@tyok 
I guess my rationale is that this still allows users to use `select` as normal just like other `Enumerable` methods. Traditionally one don't associate a hash input with `select` (but would associate `where` with a hash input). WDYT?

tyok commented:
@lulalala I do like the familiarity with ActiveRecord interface and the signature consistency between `delete`, `add`, `added?`, and `where`.

I just think overriding `select` allows you to expand the capability seamlessly. It can be implemented by detecting whether the user passed some arguments or a block:

```ruby
errors.select { |error| error.attribute == :name }   # behaves as expected
errors.select(:name, :foo, bar: 3)                   # behaves like a shortcut
errors.select(:name) { |error| error.type != :foo  } # raise an error
```

The upside is we have one method for both simple matching and complex selection logic. Complementary methods come naturally: we can simply override `reject`, `find`, `[]`, etc to accept a similar set of args.

The downside is there may be surprising and unexpected behaviors since we are redefining common enumerable methods.

Your thoughts?

lulalala commented:
@tyok I am a little hesitant as this might add confusion, but since the design is more OO now, it is easy to create extension gems for this. 

Also you mentioned that `reject` can then be easy, which I suspect is not the case (seeing the implementation of `reject` and `select` are independent of each other). I am also tight on free time unfortunately.

I am not 100% against this though. Maybe open another issue & PR to discuss this?

lulalala commented:
@tenderlove just wondering if this can be merged early? I have a feeling @rafaelfranca is overloaded with other things, since it has been a month from the last ping.

ksec commented:
Just wanted to thank @lulalala for his work on this PR. I hope @tenderlove and @rafaelfranca will have time to merge this soon.

wildjcrt commented:
+1 hope to merge!

MitsunChieh commented:
++

rafaelfranca commented:
Hi!

Just to give you an updated I'm testing this branch in our application to double check if there is no major change. So far I don't see any big change. After I confirm it works in our app, and probably on GitHub (cc @jhawthorn), I'll merge this.

jhawthorn commented on activemodel/lib/active_model/errors.rb:
```diff
@@ -509,46 +585,116 @@ def generate_message(attribute, type = :invalid, options = {})
+
+      Array(value).each do |message|
+        @errors.add(attribute, message)
+      end
```
If there are existing errors on `attribute`, this will just append to them. This feels like a behaviour change?

Maybe we should have `@errors.add(attribute)` before this loop?

lulalala replied:
I agree, but there exists this test:

```ruby
  test "values returns an array of messages" do
    errors = ActiveModel::Errors.new(self)
    errors.messages[:foo] = "omg"
    errors.messages[:baz] = "zomg"

    assert_equal ["omg", "zomg"], errors.values
  end
```

Weird, I know O_O

lulalala replied:
So this is a little complicated. The `errors[]=` was deprecated in 5.0, and removed in 5.1.

In 5.0, it actually `appended` the message: https://api.rubyonrails.org/v5.0.0/classes/ActiveModel/Errors.html#method-i-5B-5D-3D

Now in 5.1, the method was removed from API, but user can bypass them by calling `errors.messages[]=` instead. It overrides the `messages` array as you mentioned. This is actually a bad thing, for example:

```ruby
errors.messages[:foo] = "zomg"
assert_equal ["zomg"], errors[:foo] # This fails because it is not wrapped in an array
```

`errors[:foo]` now does not return an array, which can break errors rendering code.

So now I feel actually appending is a better solution, especially when we deliberately removed the `[]=` API. WDYT?

rafaelfranca commented on activemodel/lib/active_model/errors.rb:
```diff
@@ -258,13 +303,30 @@ def as_json(options = nil)
+    end
+
+    def details
+      hash = {}
```
I believe this hash needs to have an empty array as default. Otherwise `foo.errors.defails[:base].any?` will now fail.

rafaelfranca commented on activemodel/lib/active_model/errors.rb:
```diff
@@ -258,13 +303,30 @@ def as_json(options = nil)
+      hash = {}
+      group_by_attribute.each do |attribute, errors|
+        hash[attribute] = errors.map { |error|
+          normalize_detail(error.type, error.options)
```
When you add an error with a message `errors.add(:base, "Foo")` before defailts returned `error: "Foo"` now it is returning `error: :invalid`. We should keep that behavior.

lulalala replied:
The new design are build around the fundamental that `type` has to be present, and be a `:symbol`. I feel this is a breaking change that is really needed, so could we keep this (since this is a major release).

rafaelfranca replied:
I get that type needs to be always present but this is changing the return of `details` without warning. I don't think one is necessary for the other and without deprecating I don't think we can just change that or this will silently break people application. Would it possible to add a new method called `type_or_message` or something similar that does return the thing `details` used to return here?

lulalala replied:
Thanks for the suggestion and I've made the change.

However how would one go about deprecate one field in the detailed hash?

rafaelfranca replied:
If we need to, what I'm not sure yet we can return a subclass of hash that shows deprecation when one key is accessed.

lulalala replied:
I will do the deprecation in a separate PR as it is less urgent and I am occupied for the next two weeks. Thanks!

rafaelfranca commented on activemodel/lib/active_model/error.rb:
```diff
@@ -0,0 +1,197 @@
+          # this way lookup_key can also be a Proc and generated dynamically.
+          type = msg
+          msg = nil
+        end
```
Should not this method return `msg` if `options[:message]` is a string? I see no reason to do i18n lookup scope if we already passed a string.

rafaelfranca replied:
This was not happening before. Because this change I'm getting `undefined method `model_name' for nil:NilClass` in my application for errors created with `Errors.new(nil)` but with errors being added as full strings without using i18n. Before I think it was not passing through the i18n layer because this exception is not raised in the current implementation.

lulalala replied:
My bad, I think you are right. I don't recall the reason but I omitted the logic in the original `normalize_message`. I've added the return if `type` is not a symbol.

lulalala committed:
Fix messages[]= does not override value

lulalala committed:
Set default array to details

maintaining behavior errors.details[:foo].any?

lulalala committed:
Spec for display original raw type in details

lulalala reviewed (commented):
@rafaelfranca I've push the changes and left a few explanations/questions. Thanks.

lulalala commented:
@rafaelfranca I think the only test failure is the guide on 2.6, and I think it is unrelated? Can you review to see if the whole PR can be merged? Thanks :)

rafaelfranca commented:
On it! I just triggered a new build in my app to see if all tests are passing.

kamipo commented on activerecord/test/cases/associations/has_many_through_associations_test.rb:
```diff
@@ -1183,7 +1187,7 @@ def test_save_bang_should_raise_exception_when_join_record_has_errors
       c = Category.new(name: "Fishing", authors: [Author.first])
       assert_raises(ActiveRecord::RecordInvalid) do
-        c.save!
+        assert_deprecated { c.save! }
```
I'm curious why the `c.save!` raises a deprecation message.

lulalala replied:
The line 1187 used `[] <<` in the validator. So during save, that validator will be triggered and raise deprecation.

lulalala commented:
@rafaelfranca I tried another two rebase repushes, finally this one passes with no missing build problems :D

rafaelfranca commented:
There are many tests broken in my app using this branch now. Before I only had 3. Not it is more than 100. Many related to translation missing.

I'm not comfortable with merging this change. It is too big, it will break many current behavior, and add many deprecations and I'm starting to doubt if the benefits overweight the problems it will cause.

I'll investigate more which part of the code are causing all those broken tests and possibly add tests to our current suite.

lulalala commented:
@rafaelfranca I've added the change to return message if it is not a Symbol, similar to existing behavior. Could you see if that fixes things? Thanks

lulalala commented:
@rafaelfranca for the 3 errors you mentioned. If you are available, could we arrange a video call (based on your timezone)? I would like to help fixing those. I want to stay positive and see if things would work out.

lulalala commented:
@rafaelfranca For the main concern that the edge cases of message rendering logic being changed slightly, the main cause was mainly due to a new `normalize_arguments` methods I added. Reverting this part would make the whole rendering business to behave identical to existing behavior. So I'll do that.

I am guessing there is no time for me to do this before the official release. If that's the case, could this be scheduled for 6.1? Also would it be possible to have a brief chat (e.g. slack)? Thanks!

lulalala committed:
Fix spec as generate_message is no longer called during validation

lulalala commented:
@rafaelfranca (apologies for another ping) I've rewrite the code, so the rendering are all done using existing `full_message` and `generate_message` methods. This means all the edge cases you mentioned should be fixed. The code size also shrunk to be less than 1000 lines. If you have time, could you try rerun this against your tests please? Thanks.

rafaelfranca commented:
Sure! Thank you for rewriting the patch. I'd also like to say that I do think this change is a good change and I'd love to merge it , I just not sure if I'd like to merge it right now close to release. I'll test it again today to see how many things are still failing and if I find any add tests cases to our test suite so we can work on it. If we need to schedule a meeting I'll let you know. Also note that you commented in the issue during a weekend and I only look on GitHub during work hours so I'm sorry it took so long to reply.

lulalala commented:
@rafaelfranca thanks, and no problem about not replying during weekends, just that I want to update the dev status. 

Can I at least request for one thing though. The lack of feedback loop was really stressing me out during the past year. I have spent 160 hours on this, and 50 more hours just trying to resolving conflicts. Could this be put on a much higher priority after 6.0 is released? Sorry to break this to you, but I feel really unfair seeing you merge other people's work within one day, while I have to spend more hours resolving conflicts.

rafaelfranca commented:
Not using this as excuse for the delay but some PRs are easier to merge than others. Yeah, sure, this PR is my priority. I'm being running it on shopify's main application and fixing all tests I see failing there to make sure this PR doesn't have any breaking change.

rafaelfranca commented:
I just tested it again and a few tests were still failing in my app. I'll bump this to 6.1 but I promise it will be the first PR I'll merge when we start the 6.1 development.

lulalala commented:
@rafaelfranca thanks heaps! Let me know what the failures are and I will try fixing those.

rafaelfranca commented:
As promised, First PR to be merged after the development of 6.1. Thank you so much for you hard work.

rafaelfranca commented on activemodel/lib/active_model/error.rb:
```diff
@@ -0,0 +1,81 @@
+      true
+    end
+
+    def strict_match?(attribute, type, **options)
```
Do we want this to be part of the public API?

rafaelfranca reviewed (commented):
@lulalala Now that this is merged, can you work on whiting a CHANGELOG entry and improving the documentation around this change?

I usually ask this to be done in the same PR but since you already waited so long to get this merged I think it is better to do a follow up.

lulalala commented:
@rafaelfranca thanks so much, as I didn't expect it to be literally the first one to merge. I'll do the followup this weekend.

Drenmi commented:
Looking forward to this improvement! ❤️ Thanks for your hard work @lulalala and @rafaelfranca. 🙇

olimart commented:
Thank you @lulalala & @rafaelfranca 👏

henrik commented:
@lulalala Wow, great work on this! Thank you.

dineshgadge commented:
This is amazing. Thanks @lulalala and @rafaelfranca

alaarab commented:
@lulalala Something nice about the previous functionality was that errors.full_messages preserved the errors at the time of the error, of the object and its children. However, if there was some data manipulation after the error was generated, the errors will not always be preserved, in which case I would need to manually track and display model errors seperately, maybe with an attr_accessor. Is there a way in the new implementation that will return me the errors at the time they arose?

lulalala commented:
@alaarab unfortunately the current design will only lazily evaluate the error messages. Would it help if you generate a copy of the full messages (calling `full_mesages` should generate a fresh hash each time it is called) right after validation, then manipulate afterwards?

alaarab commented:
@lulalala It feels like that's the route I must go, and to store those errors in an attr_accessor variable on the model so it can be accessed to display errors from on my forms. It just feels a 'wrong' to be displaying messages from a custom attribute instead of using the object.errors directly. 

In my app, I run some manipulation in my controllers after the user submits but before the object runs .save, to limit the amount of saves in the app. If the form fails to submit, I must restore the object to its original state before that manipulation - this is where I lose some of the errors with the rails 6.1 approach. I would have preferred to still preserve the old functionality while maintaining the new functionality, like if a helper existed that replicate the old functionality as well. However, if this is not the design direction of Rails, I understand and can brew up my own solution. It just adds to the clunk when doing anything more complicated than Rails boilerplate.

thefaded commented:
@lulalala, hey, can you please help with nested attributes? Because I can't really reproduce this in Rails 6.1
Consider I have FormObject (< 6.1):

```ruby
class FormObject
  include ActiveModel::Model
  
  attr_accessor :products
  
  validates :all_products_are_valid
  
  def products_attributes=(params)
     @products = []
     
     products_attributes.each { |p| @products.push(AnotherFormObject.new(p) }
  end
  
  private
  
  def all_products_are_valid
    products.each.with_index do |product, index|
      next if product.valid?

      # Add row_id, to identify row correctly
      product.errors.messages[:row_id] = index

      errors[:products] << product.errors.messages
    end
  end
end
```
These will produce error object like this:
```
  {
    products: [
      { supplier_ref: "can't be blank", row_id: 0 },
      { quantity: "must be greater than 0", row_id: 1 }
    ]
  }
```

But with new approach I can't achieve this at all, I tried these things:
1. If the just upgrade to 6.1 without changing code above:
```
{ "products":["is invalid"] }
```
2. If use import, as this:
```ruby
  def all_products_are_valid
    products.each.with_index do |product, index|
      next if product.valid?
      
      product.errors.add(:row_id, index)
      
      product.errors.each do |error|
        errors.import(error, { attribute: "products" })
      end
    end
  end
```
it returns:
```
"products":["must be greater than 0",0,"can't be blank","must be greater than 0",1]
```
why it's plain array and keys are missing?

lulalala commented:
@thefaded Hi there! Sorry for the late reply.

The reason is that you tried to add non-message information into the messages hash, so Rails treats it as a message in 6.1. If anything it suits better in `details` hash.

Based on your use case, I think the most Railsey way would be to utilize [`index_errors`](https://bigbinary.com/blog/errors-can-be-indexed-with-nested-attrbutes-in-rails-5)

If you really want to roll your own solution, I would suggest to add the index into the error `options`, e.g.:

```ruby
  def all_products_are_valid
    products.each.with_index do |product, index|
      next if product.valid?
      
      product.errors.each do |product_error|
        product_error.options[:row_id] = index
        errors.import(product_error, { attribute: "products" })
      end
    end
  end
```

When you output the errors, you would have write code to output the error details and messages together. I would monkey-patch like this:

```
module ActiveModel
  class Errors
    def details_with_message
      hash = group_by_attribute.transform_values do |errors|
        errors.map do |error|
          error.details.tap do |details|
            details[:message] = error.message
          end
        end
      end
    end
  end
end
```

The new `details_with_message` method would then output all the things and you can get the `message` and `row_id` as wish.

lulalala commented:
@alaarab

Have you considered monkey patching? You can override the `Error#initialize` class, generate the message when called and store that as an instance variable. You then have to override `message` method to just return that message. Fingers crossed that this may break something else, but should still be fixable.

alaarab commented:
@lulalala I appreciate your response. I have already handled this in my use case without monkey patching. My solution was to create a module in app/models/concerns called last_validatable.rb:

```
 # This module adds error messages directly after validation to an attr_accessor on the model
module LastValidatable
  extend ActiveSupport::Concern

  included do
    attr_accessor :last_full_error_messages
    after_validation :populate_last_full_error_messages

    def populate_last_full_error_messages
      self.last_full_error_messages = self.errors.full_messages
    end
  end
end
```

Then on my models where I needed this functionality I simply included them with the following:

`include LastValidatable`

And finally, from my views I use this attr_accessor to display my errors from:


```
  <%= f.error_notification %>
  <%= f.error_notification message: f.object.last_full_error_messages.uniq.join("<br/>") if f.object.last_full_error_messages.present? %>
```

With this solution, I know I am not destroying any other functionality, just adding one step in my callback chain which I believe I am comfortable with at this time.  

Thank you again for responding.

bigforcegun commented:
@lulalala 

Can you please  provide an extended example or documentation how to use `import` method?

What shall I do if I need nested errors structure like this?

```
{
   "title": level_1_error,
   "items": {
      "item_1": {level_2_error}
   }
}
```

## Rails 6.0

```ruby
class NestedPerson
  # Required dependency for ActiveModel::Errors
  extend ActiveModel::Naming

  def initialize
    @errors = ActiveModel::Errors.new(self)
  end

  attr_accessor :items
  attr_reader :errors

  def validate!
    errors.add(:items, {"item_1": "some_error"})
    errors.add(:items, {"item_2": "some_error"})
  end
end
```

```bash
[2] pry(main)> q = NestedPerson.new
=> #<NestedPerson:0x000056385c816b50 @errors=#<ActiveModel::Errors:0x000056385c816ab0 @base=#<NestedPerson:0x000056385c816b50 ...>, @details={}, @messages={}>>
[3] pry(main)> q.validate!
=> [{:item_1=>"some_error"}, {:item_2=>"some_error"}]
[4] pry(main)> q.errors
=> #<ActiveModel::Errors:0x000056385c816ab0
 @base=#<NestedPerson:0x000056385c816b50 @errors=#<ActiveModel::Errors:0x000056385c816ab0 ...>>,
 @details={:items=>[{:error=>{:item_1=>"some_error"}}, {:error=>{:item_2=>"some_error"}}]},
 @messages={:items=>[{:item_1=>"some_error"}, {:item_2=>"some_error"}]}>
[5] pry(main)> q.errors.as_json
=> {:items=>[{:item_1=>"some_error"}, {:item_2=>"some_error"}]}
[6] pry(main)>
```
## Rails 6.1

```ruby
class NestedPerson
  # Required dependency for ActiveModel::Errors
  extend ActiveModel::Naming

  def initialize
    @errors = ActiveModel::Errors.new(self)
  end

  attr_accessor :items
  attr_reader :errors

  def validate!
    errors.add(:items, { "item_1": "some_error" })
    errors.add(:items, { "item_2": "some_error" })
  end

  # The following methods are needed to be minimally implemented

  def read_attribute_for_validation(attr)
    send(attr)
  end

  def self.human_attribute_name(attr, options = {})
    attr
  end

  def self.lookup_ancestors
    [self]
  end

end
```
```bash
[7] pry(main)> q = NestedPerson.new
=> #<NestedPerson:0x0000564c806e4128 @errors=#<ActiveModel::Errors:0x0000564c806e4100 @base=#<NestedPerson:0x0000564c806e4128 ...>, @errors=[]>>
[8] pry(main)> q.validate!
=> #<ActiveModel::Error attribute=items, type=invalid, options={:item_2=>"some_error"}>
[9] pry(main)> q.errors
=> #<ActiveModel::Errors:0x0000564c806e4100
 @base=#<NestedPerson:0x0000564c806e4128 @errors=#<ActiveModel::Errors:0x0000564c806e4100 ...>>,
 @errors=
  [#<ActiveModel::Error attribute=items, type=invalid, options={:item_1=>"some_error"}>,
   #<ActiveModel::Error attribute=items, type=invalid, options={:item_2=>"some_error"}>]>
[10] pry(main)> q.errors.as_json
=> {:items=>["is invalid", "is invalid"]}
```

lulalala commented:
@bigforcegun 

I think you are trying similar things, wanting to add multiple errors to the `has_many` association with index.

Could you read https://github.com/rails/rails/pull/32313#issuecomment-776406281 and see if that helps? Thanks!

akostadinov commented:
> Can we deprecate to_xml? I bet is rarely used, and can exists as a gem.

Nobody has an XML API that would return errors? Why need to use a separate gem?

rafaelfranca commented:
Because we don't want to maintain something that Rails itself doesn't promote. In the past, we used to promote XML APIs. Those are long gone. We still support you to implement a XML API if you want to, but we will not maintain code to make that happen.

akostadinov commented:
I don't understand what does it mean to support a use case without maintaining code to make that happen. Also not sure who that "we" is and whether any formal decisions were made to prevent somebody to re-introduce the `to_xml` (which is quiet simple as far as I see).

Either way, I assume new projects supporting XML API are a very rare occasion but for sure a lot of apps exist that need to workaround simple things like this when upgrading Rails. In my opinion a respectable app would not be able to drop a whole API without loosing user base. And something so fundamental that was promoted in the past being removed from the framework.. idk, looks wrong to me.

rafaelfranca commented:
"We" is the Rails core team, and yes, formal decisions to prevent people to re-introduce `to_xml` where made.

When I say that "We still support" I mean that Rails still supports that. If you want to generate XML representation of a `ActiveModel::Error`, that like you said it is quite simple, you can do that in your application, or in a library. Rails will not do it. XML API support was removed in Rails 4.0.

If you want support to `to_xml` in a `ActiveMovel::Error` I think I'd be ok to add to https://github.com/rails/activemodel-serializers-xml, just not to Rails itself.
